bindings/cs/Protean-cs/Protean.Net/BinaryWriter.cs

```
@@ -65,357 +65,561 @@ public static byte[] ToBytes(Variant value, BinaryMode mode)
-		}
-
-		protected void Write(float arg)
-		{
-			Write(BitConverter.GetBytes(arg));
-		}
-
-		protected void Write(double arg)
-		{
-			Write(BitConverter.DoubleToInt64Bits(arg));
-		}
-
-		private void Write(byte[] arg)
-		{
-			m_filter.Write(arg, 0, arg.Length);
-		}
-
-		protected void Write(bool arg)
-		{
-			Write((arg ? 1 : 0));
-		}
-
-		protected void Write(byte[] bytes, bool writePadding)
-		{
-			Write(bytes);
-
-			if (writePadding)
-			{
-				int residual = (4 - (bytes.Length % 4)) % 4;
-				for (int i = 0; i < residual; ++i)
-				{
-					Write((byte)0);
-				}
-			}
-		}
-
-	    protected void Write(DataTable arg)
-	    {
+        }
+
+        public static byte[] ToBytes(Variant value)
+        {
+            return ToBytes(value, BinaryMode.Default);
+        }
+
+        public void Write(Variant v)
+        {
+            WriteHeader();
+            WriteVariant(v);
+        }
+
+        public static int GetBytes(Variant value, BinaryMode mode, byte[] buffer, int bufferIndex)
+        {
+            using (MemoryStream ms = new MemoryStream(buffer, bufferIndex, buffer.Length - bufferIndex))
+            {
+                using (BinaryWriter writer = new BinaryWriter(ms, mode))
+                {
+                    writer.Write(value);
+                }
+
+                return (int)ms.Position;
+            }
+        }
+
+        public static int GetByteCount(Variant variant)
+        {
+            return 12 + GetByteCountVariant(variant);
+        }
+
+        private void WriteHeader()
+        {
+            // Create and write the Protean header
+            m_stream.Write(BitConverter.GetBytes(BinaryConstants.PROTEAN_MAGIC), 0, 4);
+            m_stream.Write(BitConverter.GetBytes(BinaryConstants.VERSION_MAJOR << 16 | BinaryConstants.VERSION_MINOR), 0, 4);
+            m_stream.Write(BitConverter.GetBytes((int)m_mode), 0, 4);
+        }
+
+        // This doesn't seem to support Unicode properly
```

Weren't you going to change this so that it will throw for an out-of-range character, rather than silently mangling it?

---

I'm a bit reluctant to change behaviour, it might break existing code